### PR TITLE
feat(kafka): aws_msk_config and aws_msk_cluster modules have been added

### PR DESCRIPTION
**Pull request by oukooveu**
##### SUMMARY
Modules for [Amazon Managed Streaming for Apache Kafka ](https://docs.aws.amazon.com/msk/latest/developerguide/what-is-msk.html)(Amazon MSK) have been added. There are two modules: `aws_msk_config` for managing MSK configuration and `aws_msk_cluster` -- for clusters.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
- aws_msk_config
- aws_msk_cluster

##### ADDITIONAL INFORMATION
All MSK cluster related operations are time consuming, for example create takes 20-30 minutes, update Kafka version -- more than one hour, update configuration -- 10-15 minutes.

Integration tests work in my own account, please let me know if I can help with an evidences.

##### REFERENCES
Related aws-terminator PR: https://github.com/mattclay/aws-terminator/pull/143
Boto3: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/kafka.html

### CODE REVIEW

File: plugins/modules/aws_msk_config.py
Comment (tremble) [inline plugins/modules/aws_msk_config.py:207]: It's probably better to do the comparison in a dedicated function, that would mean you could simplify things a little:

```
def _is_changed(current, module):
    prop_module = {}
    for k, v in module.params.get("config").items():
        prop_module[str(k)] = str(v)
    if prop_to_dict(response.get("ServerProperties", "")) != prop_module:
        return True
    if response.get("Description", "") != module.params.get("description"):
        return True
    return False
```

```
        if not _is_changed(response, module):
            return False, response
```

it makes the code a little easier to follow

File: plugins/modules/aws_msk_config.py
Comment (tremble) [inline plugins/modules/aws_msk_config.py:227]: Please wrap in try/except and add retries

File: tests/integration/targets/aws_msk_config/aliases
Comment (tremble) [inline tests/integration/targets/aws_msk_config/aliases:2]: Assuming it'll be possible to run this in CI we should wait for the updated permissions before merging.

**Review by tremble — CHANGES_REQUESTED** [review]
I've not had chance to go through this 100%

Here's some initial feedback.

**Comment by tremble** [comment]
Thanks for taking the time to open this PR.  If you need help (or have more questions about our CI) feel free to reach out on #ansible-aws again.

File: plugins/modules/aws_msk_config.py
Comment (oukooveu) [reply plugins/modules/aws_msk_config.py:63]: it's just a misprint due to copy/paste, thanks for notice

File: plugins/modules/aws_msk_config.py
Comment (oukooveu) [reply plugins/modules/aws_msk_config.py:132]: and the same for a second module i suppose, done

File: plugins/modules/aws_msk_config.py
Comment (oukooveu) [reply plugins/modules/aws_msk_config.py]: fixed

File: plugins/modules/aws_msk_config.py
Comment (oukooveu) [reply plugins/modules/aws_msk_config.py:182]: should be fixed

File: plugins/modules/aws_msk_config.py
Comment (oukooveu) [reply plugins/modules/aws_msk_config.py]: thanks, done for all aws related (non-paginated) calls

File: plugins/modules/aws_msk_config.py
Comment (oukooveu) [reply plugins/modules/aws_msk_config.py:205]: Yes, it's for correct comparison two dictionaries below.

File: plugins/modules/aws_msk_config.py
Comment (oukooveu) [reply plugins/modules/aws_msk_config.py:215]: fixed

File: plugins/modules/aws_msk_config.py
Comment (oukooveu) [reply plugins/modules/aws_msk_config.py]: fixed and moved to a dedicated function

File: plugins/modules/aws_msk_config.py
Comment (oukooveu) [reply plugins/modules/aws_msk_config.py]: fixed

File: plugins/modules/aws_msk_config.py
Comment (oukooveu) [reply plugins/modules/aws_msk_config.py:216]: agree, done, thanks for suggestion

File: tests/integration/targets/aws_msk_config/aliases
Comment (oukooveu) [reply tests/integration/targets/aws_msk_config/aliases:2]: There is an another (not only permissions) "issue" with tests execution time, for `aws_msk_config` they are pretty fast:
```
time ansible-test integration aws_msk_config --allow-unsupported --docker
...
PLAY RECAP *********************************************************************
testhost                   : ok=22   changed=6    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   

AWS ACTIONS: ['kafka:CreateConfiguration', 'kafka:DeleteConfiguration', 'kafka:DescribeConfigurationRevision', 'kafka:ListConfigurations', 'kafka:UpdateConfiguration']
Emulate Docker CLI using podman. Create /etc/containers/nodocker to quiet msg.

real	0m54.458s
user	0m2.889s
sys	0m1.704s
```
but for `aws_msk_cluster` consumes a lot of time:
```
time ansible-test integration aws_msk_cluster --allow-unsupported --docker
...
PLAY RECAP *********************************************************************
testhost                   : ok=34   changed=12   unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   

AWS ACTIONS: ['ec2:CreateSubnet', 'ec2:CreateTags', 'ec2:CreateVpc', 'ec2:DeleteSubnet', 'ec2:DeleteVpc', 'ec2:DescribeAvailabilityZones', 'ec2:DescribeSubnets', 'ec2:DescribeTags', 'ec2:DescribeVpcAttribute', 'ec2:DescribeVpcClassicLink', 'ec2:DescribeVpcs', 'ec2:ModifyVpcAttribute', 'kafka:CreateCluster', 'kafka:CreateConfiguration', 'kafka:DeleteCluster', 'kafka:DeleteConfiguration', 'kafka:DescribeCluster', 'kafka:DescribeConfigurationRevision', 'kafka:GetBootstrapBrokers', 'kafka:ListClusters', 'kafka:ListConfigurations', 'kafka:ListTagsForResource', 'kafka:TagResource', 'kafka:UntagResource', 'kafka:UpdateBrokerStorage', 'kafka:UpdateMonitoring']
Emulate Docker CLI using podman. Create /etc/containers/nodocker to quiet msg.

real	45m51.672s
user	0m3.401s
sys	0m1.888s
```

**Comment by jillr** [comment]
MSK creation and updates are quite slow, the test suite took around 30 minutes to run for me, we might want to mark these tests unstable so they only run when there are changes to the modules.

**Comment by jillr** [comment]
recheck

**Comment by oukooveu** [comment]
tests (build-ansible-collection) have been failed with error:
```
ERROR! Unexpected Exception, this is probably a bug: cannot import name 'CollectionRequirement'
the full traceback was:

Traceback (most recent call last):
  File "/home/zuul/src/github.com/ansible-network/releases/.tox/venv/bin/ansible-galaxy", line 92, in <module>
    mycli = getattr(__import__("ansible.cli.%s" % sub, fromlist=[myclass]), myclass)
  File "/home/zuul/src/github.com/ansible-network/releases/.tox/venv/lib/python3.6/site-packages/ansible/cli/galaxy.py", line 24, in <module>
    from ansible.galaxy.collection import (
ImportError: cannot import name 'CollectionRequirement'
```

**Comment by alinabuzachis** [comment]
recheck

**Comment by alinabuzachis** [comment]
recheck

**Comment by oukooveu** [comment]
recheck

**Comment by alinabuzachis** [comment]
@oukooveu Thank you for working on these new modules. Could you please rebase your PR?

**Comment by oukooveu** [comment]
> @oukooveu Thank you for working on these new modules. Could you please rebase your PR?

done

File: plugins/modules/aws_msk_cluster.py
Comment (tremble) [inline plugins/modules/aws_msk_cluster.py:24]:
```suggestion
        description: Create (C(present)) or delete (C(absent)) cluster.
```

File: plugins/modules/aws_msk_cluster.py
Comment (tremble) [inline plugins/modules/aws_msk_cluster.py:73]:
```suggestion
            - Client subnets can't be in Availability Zone us-east-1e.
```
For any other reviewers, this is an AWS restriction and they only list us-east-1e
https://docs.amazonaws.cn/en_us/msk/1.0/apireference/clusters.html

File: plugins/modules/aws_msk_cluster.py
Comment (tremble) [inline plugins/modules/aws_msk_cluster.py:201]:
```suggestion
        description: How many seconds to wait. Cluster creation can take up to 20-30 minutes.
```

File: plugins/modules/aws_msk_cluster.py
Comment (tremble) [inline plugins/modules/aws_msk_cluster.py:297]: While the _with_backup calls will automatically retry, they'll let non-temporary exceptions fall through.  BotoCoreException and ClientException should be caught here to avoid spitting big traces out at users.

File: plugins/modules/aws_msk_cluster.py
Comment (tremble) [inline plugins/modules/aws_msk_cluster.py:329]:
```suggestion
def wait_for_cluster_state(client, module, arn, state="ACTIVE"):
    # As of 2021-06 boto3 doesn't offer any built in waiters
```

**Review by tremble — APPROVED** [review]
Generally looks good, with some minor niggles.  It'd be good if someone else can take a look I've not had the cycles for a thorough review.

File: plugins/modules/aws_msk_cluster.py
Comment (oukooveu) [reply plugins/modules/aws_msk_cluster.py]: Fixed

**Comment by goneri** [comment]
recheck

**Comment by goneri** [comment]
rec

**Comment by oukooveu** [comment]
@jillr are you going to execute tests by yourself or do we need to remove "unsupported" tags from tests?

**Comment by goneri** [comment]
If the tests can be run in the CI, please remove the `unsupported` flag.

**Comment by oukooveu** [comment]
> If the tests can be run in the CI, please remove the `unsupported` flag.

I'm not sure required policies were applied to target account, waiting for @jillr for confirmation.

**Comment by jillr** [comment]
@oukooveu Yes, your terminator PR was merged and the new policies were deployed to our CI account

**Comment by oukooveu** [comment]
@jillr I've removed `unsupported` from aliases and tests were passed successfully but actually integration tests were not started and `ansible-test: Run the test suite` task was skipped in all three `ansible-test-cloud-integration-aws-py36` batches. Don't understand why, the condition is:
```
- name: Run the test suite
  args:
    chdir: "{{ _test_location }}"
    executable: /bin/bash
  environment: "{{ ansible_test_environment | default({}) }}"
  shell: "source {{ ansible_test_venv_path }}/bin/activate; {{ ansible_test_executable }} {{ ansible_test_test_command }} {{ ansible_test_options }} {{ ansible_test_verbosity }} {{ _integration_targets }}"
  when: (ansible_test_changed|bool == False) or (ansible_test_changed|bool and _integration_targets|length > 0)
```
and `_integration_targets` contains changed modules as we can see in job output:
image
also `ansible_test_changed` should be `true` because of definition in `zuul.d/ansible-cloud-jobs.yaml`:
```
- job:
    name: ansible-test-cloud-integration-aws-py36
    parent: ansible-core-ci-aws-session
    nodeset: controller-python36
    dependencies:
      - name: build-ansible-collection
    pre-run:
      - playbooks/ansible-test-base/pre.yaml
      - playbooks/ansible-cloud/aws/pre.yaml
    run: playbooks/ansible-test-base/run.yaml
    # TODO: disabled for now, required for Ara
    # post-run: playbooks/ansible-test-base/post.yaml
    required-projects:
      - name: github.com/ansible/ansible
    timeout: 3600
    vars:
      ansible_test_collections: true
      ansible_collections_repo: "{{ zuul.project.canonical_name }}"
      ansible_test_command: integration
      ansible_test_enable_ara: false
      ansible_test_python: 3.6
      ansible_test_changed: true
      ansible_test_split_in: 3
```
Could you or @goneri take a look please?

**Comment by oukooveu** [comment]
I've found what's wrong with tests run, they are splitting in incorrect way, the issue is in `roles/ansible-test/tasks/split_targets.yaml`:
```
- set_fact:
    _start_at: "{{ number_entries|int // ansible_test_split_in * (ansible_test_do_number -1)|int }}"
    _end_at: "{{ (number_entries|int // ansible_test_split_in * (ansible_test_do_number |int)) }}"
```
and when `ansible_test_split_in` is greater than `number_entries` we get empty list.

Possible fix (if you need even distribution):
```
--- a/roles/ansible-test/tasks/split_targets.yaml
+++ b/roles/ansible-test/tasks/split_targets.yaml
@@ -10,9 +10,10 @@
 - set_fact:
     number_entries: "{{ ansible_test_targets.stdout_lines|length }}"
 - set_fact:
-    _start_at: "{{ number_entries|int // ansible_test_split_in * (ansible_test_do_number -1)|int }}"
-    _end_at: "{{ (number_entries|int // ansible_test_split_in * (ansible_test_do_number |int)) }}"
-- debug:
-    msg: "Do roles: {{ ansible_test_targets.stdout_lines[_start_at|int:_end_at|int] }}"
+    _integration_targets: []
 - set_fact:
-    _integration_targets: "{{ ansible_test_targets.stdout_lines[_start_at|int:_end_at|int]|join(' ') }}"
+    _integration_targets: "{{ _integration_targets + [ ansible_test_targets.stdout_lines[item] ] }}"
+  when: item % ansible_test_split_in == ansible_test_do_number|int - 1
+  loop: "{{ range(0,number_entries|int)|list }}"
+- debug:
+    msg: "Do roles: {{ _integration_targets }}"

```

**Comment by oukooveu** [comment]
recheck

**Comment by oukooveu** [comment]
PR with fix for zuul jobs distribution: https://github.com/ansible/ansible-zuul-jobs/pull/968 (added as dependency).

**Comment by oukooveu** [comment]
`aws_msk_config` tests are okay, `aws_msk_cluster` tests are failed with:
```
2021-06-26 13:48:38.225976 | centos-8 | TASK [aws_msk_cluster : create msk cluster (check mode)] ***********************
2021-06-26 13:48:38.226082 | centos-8 | task path: /home/zuul/.ansible/collections/ansible_collections/community/aws/tests/integration/targets/aws_msk_cluster/tasks/test_create.yml:2
2021-06-26 13:48:38.258133 | centos-8 | <testhost> ESTABLISH LOCAL CONNECTION FOR USER: zuul
2021-06-26 13:48:38.258388 | centos-8 | <testhost> EXEC /bin/sh -c 'echo ~zuul && sleep 0'
2021-06-26 13:48:38.263996 | centos-8 | <testhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/zuul/.ansible/tmp `"&& mkdir "` echo /home/zuul/.ansible/tmp/ansible-tmp-1624715318.2630663-4289-87016496591224 `" && echo ansible-tmp-1624715318.2630663-4289-87016496591224="` echo /home/zuul/.ansible/tmp/ansible-tmp-1624715318.2630663-4289-87016496591224 `" ) && sleep 0'
2021-06-26 13:48:38.398708 | centos-8 | Using module file /home/zuul/.ansible/collections/ansible_collections/community/aws/plugins/modules/aws_msk_cluster.py
2021-06-26 13:48:38.399639 | centos-8 | <testhost> PUT /home/zuul/.ansible/tmp/ansible-local-4087ta6jc607/tmpd7mtwhki TO /home/zuul/.ansible/tmp/ansible-tmp-1624715318.2630663-4289-87016496591224/AnsiballZ_aws_msk_cluster.py
2021-06-26 13:48:38.400139 | centos-8 | <testhost> EXEC /bin/sh -c 'chmod u+x /home/zuul/.ansible/tmp/ansible-tmp-1624715318.2630663-4289-87016496591224/ /home/zuul/.ansible/tmp/ansible-tmp-1624715318.2630663-4289-87016496591224/AnsiballZ_aws_msk_cluster.py && sleep 0'
2021-06-26 13:48:38.406887 | centos-8 | <testhost> EXEC /bin/sh -c 'ANSIBLE_DEBUG_BOTOCORE_LOGS=True /home/zuul/venv/bin/python3.6 /home/zuul/.ansible/tmp/ansible-tmp-1624715318.2630663-4289-87016496591224/AnsiballZ_aws_msk_cluster.py && sleep 0'
2021-06-26 13:48:39.092279 | centos-8 | <testhost> EXEC /bin/sh -c 'rm -f -r /home/zuul/.ansible/tmp/ansible-tmp-1624715318.2630663-4289-87016496591224/ > /dev/null 2>&1 && sleep 0'
2021-06-26 13:48:39.101776 | centos-8 | The full traceback is:
2021-06-26 13:48:39.101799 | centos-8 | Traceback (most recent call last):
2021-06-26 13:48:39.101807 | centos-8 |   File "/tmp/ansible_aws_msk_cluster_payload_5j4bpvgy/ansible_aws_msk_cluster_payload.zip/ansible_collections/community/aws/plugins/modules/aws_msk_cluster.py", line 298, in find_cluster_by_name
2021-06-26 13:48:39.101814 | centos-8 |   File "/tmp/ansible_aws_msk_cluster_payload_5j4bpvgy/ansible_aws_msk_cluster_payload.zip/ansible_collections/amazon/aws/plugins/module_utils/cloud.py", line 154, in retry_func
2021-06-26 13:48:39.101821 | centos-8 |     raise e
2021-06-26 13:48:39.101828 | centos-8 |   File "/tmp/ansible_aws_msk_cluster_payload_5j4bpvgy/ansible_aws_msk_cluster_payload.zip/ansible_collections/amazon/aws/plugins/module_utils/cloud.py", line 144, in retry_func
2021-06-26 13:48:39.101834 | centos-8 |     return f(*args, **kwargs)
2021-06-26 13:48:39.101840 | centos-8 |   File "/tmp/ansible_aws_msk_cluster_payload_5j4bpvgy/ansible_aws_msk_cluster_payload.zip/ansible_collections/community/aws/plugins/modules/aws_msk_cluster.py", line 287, in list_clusters_with_backoff
2021-06-26 13:48:39.101846 | centos-8 |   File "/home/zuul/venv/lib/python3.6/site-packages/botocore/paginate.py", line 449, in build_full_result
2021-06-26 13:48:39.101853 | centos-8 |     for response in self:
2021-06-26 13:48:39.101859 | centos-8 |   File "/home/zuul/venv/lib/python3.6/site-packages/botocore/paginate.py", line 255, in __iter__
2021-06-26 13:48:39.101865 | centos-8 |     response = self._make_request(current_kwargs)
2021-06-26 13:48:39.101872 | centos-8 |   File "/home/zuul/venv/lib/python3.6/site-packages/botocore/paginate.py", line 332, in _make_request
2021-06-26 13:48:39.101878 | centos-8 |     return self._method(**current_kwargs)
2021-06-26 13:48:39.101886 | centos-8 |   File "/home/zuul/venv/lib/python3.6/site-packages/botocore/client.py", line 386, in _api_call
2021-06-26 13:48:39.101900 | centos-8 |     return self._make_api_call(operation_name, kwargs)
2021-06-26 13:48:39.101906 | centos-8 |   File "/home/zuul/venv/lib/python3.6/site-packages/botocore/client.py", line 705, in _make_api_call
2021-06-26 13:48:39.101911 | centos-8 |     raise error_class(parsed_response, operation_name)
2021-06-26 13:48:39.101917 | centos-8 | botocore.errorfactory.BadRequestException: An error occurred (BadRequestException) when calling the ListClusters operation: The resource name contains too many or too few characters
```

**Review by alinabuzachis — APPROVED** [review]
@oukooveu Thank you for working on this. LGTM.

**Review by goneri — APPROVED** [review]

**Comment by goneri** [comment]
Thank you @oukooveu!

**Pull request merged**
